File: src/leetcode/python/medium/ConvertSortedArrayToBinarySearchTree_108.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ConvertSortedArrayToBinarySearchTree_108(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        length = len(nums)
        if length == 0: return None
        mid = length / 2
        head = TreeNode(nums[mid])
        head.left = self.leaf_node(nums, 0, mid)
        head.right = self.leaf_node(nums, mid + 1, length)
        return head

    def leaf_node(self, nums, i, j):
        if i == j: return None
        mid = (i + j) / 2
        leaf = TreeNode(nums[mid])
        leaf.left = self.leaf_node(nums, i, mid)
        leaf.right = self.leaf_node(nums, mid + 1, j)
        return leaf

File: src/leetcode/java/ConvertSortedArrayToBinarySearchTree_108.java
package leetcode.java;

public class ConvertSortedArrayToBinarySearchTree_108 {
  public TreeNode sortedArrayToBST(int[] nums) {
    if (nums.length == 0) return null;
    TreeNode head = new TreeNode(nums[nums.length / 2]);
    head.left = getLeafNode(nums, 0, nums.length / 2);
    head.right = getLeafNode(nums, nums.length / 2 + 1, nums.length);
    return head;
  }

  private TreeNode getLeafNode(int[] nums, int i, int j) {
    if (i == j) return null;
    int mid = (i + j) / 2;
    TreeNode leaf = new TreeNode(nums[mid]);
    leaf.left = getLeafNode(nums, i, mid);
    leaf.right = getLeafNode(nums, mid + 1, j);
    return leaf;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
